Add namespace metadata configuration document (#901)
giantswarm/docs: src/content/app-platform/namespace-configuration/index.md

---
linkTitle: App target namespace metadata
title: How to annotate an app's target namespace via its App CR
description: Configuring metadata for the target namespace of an app via its app CR so it can be used by other tools such as service meshes.

> weight: 20

Reviewer: ```suggestion
weight: 40
```

This controls the order on https://docs.giantswarm.io/app-platform/ 

This feature is great but only applies to some app CRs. Can you also bump `apps` to 50?

Author: Thanks, TIL this is ordering by descending order.

---
linkTitle: App namespace metadata

> title: App's namespace metadata

Reviewer 2: ```suggestion
title: How to annotate an app's namespace via its App CR
```

Reviewer 2: Want to make it clearer in the headline what this helps the user to do.

description: Configuring metadata for the target namespace of an app via its app CR so it can be used by other tools such as service meshes.
weight: 20
menu:
  main:
    parent: app-platform
owner:
  - https://github.com/orgs/giantswarm/teams/team-batman

> aliases:
>   - /reference/app-configuration/

Reviewer: ```suggestion
```

Author: 🙈 Sorry, it was left over from copy & paste.

user_questions:
  - How can I label the app CR's namespace?
  - How can I annotate the app CR's namespace?
---

> # App CR's namespace configuration

Author: To follow the defaulting validation logic document, I think it makes sense to add it as below. 
```
# App CR's target namespace configuration

## Overview
```
Right?

> # App CR's namespace configuration

Reviewer: Yes I missed that. Thanks good to keep aligned with the other page. 👍

> At some point, you might want to put the specific labels or annotations to the app's namespace to bring benefits from `linkerd` or `kiam`.

Reviewer 2: ```suggestion
At some point, you might want to put the specific labels or annotations to the app's namespace to bring benefits from `linkerd` or `kiam`. For example, to include everything in a namespace to the Linkerd service mesh. Or, to tell KIAM you do allow Loki to get logs about pods running in this namespace. 
```

Maybe add example to illustrate? Is this correct?

Author: Perfect, thanks.

This feature enables you to add labels and annotations to the app's namespace by updating [App CRs]({{< relref "/ui-api/management-api/crd/apps.application.giantswarm.io.md" >}}).

> ## Configuring the App CR's namespace
> 
> ### Updating labels/annotations

Reviewer: ```suggestion
## Configuring labels / annotations
```

I think we can simplify to have less headings.

> At some point, you might want to put the specific labels or annotations to the app's namespace to bring benefits from `linkerd` or `kiam`. For example, to include everything in a namespace to the Linkerd service mesh. Or, to tell KIAM you do allow Loki to get logs about pods running in this namespace.

Reviewer: ```suggestion
At some point, you might want to add specific labels or annotations to an app's target namespace to enable other tools such as `linkerd` or `kiam`. For example, to include all pods in a namespace in the `linkerd` service mesh. Or, to tell `kiam` you do allow Loki to get logs from pods running in this namespace.
```

> This feature enables you to add labels and annotations to the app's namespace by updating [App CRs]({{< relref "/ui-api/management-api/crd/apps.application.giantswarm.io.md" >}}).

Reviewer: ```suggestion
This feature enables you to add labels and annotations to the app's target namespace via its [App CR]({{< relref "/ui-api/management-api/crd/apps.application.giantswarm.io.md" >}}). This is simpler than using other solutions such as triggering a Kubernetes jpb via a Helm hook.
```

I think we need to describe the alternative so its clear why we built this.

> You can use `.spec.namespaceConfig` to configure the namespace metadata. You can either choose `.spec.namespaceConfig.annotations` or
> `spec.namespaceConfig.labels` and provide the values as a key/value map. App platform then ensures the namespace exists and has this metadata.

Reviewer: ```suggestion
You can use `.spec.namespaceConfig` to configure the namespace metadata. You can set `.spec.namespaceConfig.annotations` and / or
`spec.namespaceConfig.labels` and provide the values as a key/value map. App platform then ensures the namespace exists and has this metadata.
```

For example, if you want to add an annotation `linked.io/inject: enabled` into the namespace `loki1`, try as below.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: loki-app
  namespace: loki0
spec:
  catalog: giantswarm
  name: loki
  namespace: loki
  namespaceConfig:
    annotations:
      linkerd.io/inject: "enabled"
```

Now you can check the namespace and the namespace `loki0` would be the same as below.

```yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    linkerd.io/inject: "enabled"
  name: giantswarm
```

If you want to update annotations in the namespace, use `spec.namespaceConfig.annotation` as below.

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: kiam-app
  namespace: kiam0
spec:
  catalog: giantswarm
  name: kiam
  namespace: kiam
  namespaceConfig:
    labels:
      monitoring: "enabled"
```

> ### Validation in App CR's Namespace Config

Author: Much better, thanks.

In [default validation logic]({{< relref "/app-platform/defaulting-validation" >}}), app-admission-controller checks whether multiple app CRs are updating the same namespace

> with different values. This logic would avoid the situation where the same namespace keeps changing at each app CR's reconciliation.

Reviewer 2: ```suggestion
with different values. If that is the case, it prompts the user to confirm which namespace to use. This logic would avoid the situation where the same namespace keeps changing at each app CR's reconciliation.
```

What does app-admission-controller do in the case where multiple app CRs are updating the same namespace w different values? I just wrote some BS that is probably wrong. Haha.

Author: It's not BS! :) I think we can say `If this is the case, it rejects entries of the latest app CR`

Author: I would like to use Ross's suggestion here, https://github.com/giantswarm/docs/pull/901#discussion_r613106463 
Using the word `prevents` makes sense to me.
